.travis.yml

```diff
@@ -90,21 +90,73 @@ install:
   - export AWESOME_VERSION="${TRAVIS_BRANCH}-g$(git rev-parse --short HEAD)"
   - 'if [ "$TRAVIS_PULL_REQUEST" != false ]; then AWESOME_VERSION="${AWESOME_VERSION}-PR${TRAVIS_PULL_REQUEST}"; fi'
 
+  # Functions for custom Travis folds.
+  - |
+    travis_fold_start() {
+        echo -n "travis_"
+        echo "fold:start:$1"
+        travis_fold_current="$1"
+    }
+    travis_fold_end() {
+        echo -n "travis_"
+        echo "fold:end:$travis_fold_current"
+    }
 script:
   - export CMAKE_ARGS="-DLUA_LIBRARY=${LUALIBRARY} -DLUA_INCLUDE_DIR=${LUAINCLUDE} -D OVERRIDE_VERSION=$AWESOME_VERSION -D DO_COVERAGE=${DO_COVERAGE}"
   - |
     if [ -n "$BUILD_IN_DIR" ]; then
-        SOURCE_DIRECTORY="$PWD"
-        mkdir "$BUILD_IN_DIR"
-        cd "$BUILD_IN_DIR"
-        cmake $CMAKE_ARGS "$SOURCE_DIRECTORY"
+      travis_fold_start "build_in_dir"
+      SOURCE_DIRECTORY="$PWD"
+      mkdir "$BUILD_IN_DIR"
+      cd "$BUILD_IN_DIR"
+      cmake $CMAKE_ARGS "$SOURCE_DIRECTORY"
+      cd -
+      travis_fold_end
+    fi
+  - |
+    travis_fold_start "make"
+    make
+    travis_fold_end
+    if [ "$TRAVIS_TEST_RESULT" = 0 ]; then
+      travis_fold_start "make.install"
+      sudo env PATH=$PATH make install && awesome --version
+      travis_fold_end
+    fi
+    if [ "$TRAVIS_TEST_RESULT" = 0 ]; then
+      travis_fold_start "make.check"
+      make check
+      travis_fold_end
+    fi
+  - |
+    if [ "$TRAVIS_TEST_RESULT" = 0 ] && [ "$DO_COVERAGE" = "true" ]; then
+      travis_fold_start "DO_COVERAGE"
+      make check-coverage || exit 1;
+      sed -i "1 i\\require('luacov.runner')('"$PWD"/.luacov')" build/awesomerc.lua || exit 1
+      BUILD_DIRECTORY="" tests/run.sh || exit 1
+      travis_fold_end
     fi
-  - make && sudo env PATH=$PATH make install && awesome --version && make check
   - |
-    if [ "$DO_COVERAGE" = "true" ]; then
-        make check-coverage || exit 1;
-        sed -i "1 i\\require('luacov.runner')('"$PWD"/.luacov')" build/awesomerc.lua || exit 1
-        BUILD_DIRECTORY="" tests/run.sh || exit 1
+    if [ "$TEST_PREV_COMMITS" = 1 ]; then
+      # Check each commit separately (to make git-bisect less annoying).
+      echo "Testing previous commits based on TRAVIS_COMMIT_RANGE=$TRAVIS_COMMIT_RANGE"
+      commits="$(git rev-list --abbrev-commit "$TRAVIS_COMMIT_RANGE")"
+      echo "Commits: $commits"
+      n="$(echo comiits | wc -l)"
+      i=0
+      failed=
+      for commit in $commits; do
+        i=$((i+1))
+        travis_fold_start "test_commit_${commit}_.$i.$n"
+        git checkout "$commit"
+        if ! make check; then
```

Comment:
I don't think just `make check` is enough. We don't have the dependencies set up correctly so that this e.g. rebuilds the `awesome` binary if needed. I guess you need to run `make all check` for this.